### DISCUSSION TITLE
Sanni reader can't dump R.C. Pro-Am cart

### DISCUSSION THREAD

**Discussion by ghost** [sanni/cartreader#397] [Need help]
I have an R.C. Pro-Am cart with the following PCB:

https://nescartdb.com/profile/view/629/rc-pro-am

The database says it uses iNES mapper 7 and PRG0 is 64kb, but the only options for that mapper are 128 or 256. I tried those and the binary that results is no good. 

How can I dump this cart?

**Comment by ghost** [comment]
Also having a similar issue with Cobra Triangle, which uses mapper 7 and 128 for PRG0.

**Comment by whitezombie2000** [comment]
Confirmed on v8.3. I choose mapper 7 and can only choose. 128 or 256 for PRG Size.

**Comment by ghost** [comment]
Got Cobra Triangle to dump after a few more tries. Also got R.C. Pro-Am dumped by using a hex editor to cut the 128kB file in half. Hopefully, a 64kB option will be added for mapper 7.

**Comment by sanni** [comment]
You can edit the mapsize[] array in NES.ino to allow different sizes. In this case the line that starts with 7 is mapper 7 and changing the 3 after the 7 to a 2 will allow 64KB progmem size.  

Before:
````
******************************************
  Supported Mappers
 *****************************************/
// Supported Mapper Array (iNES Mapper #s)
// Format = {mapper,prglo,prghi,chrlo,chrhi,ramlo,ramhi}
static const byte PROGMEM mapsize [] = {
  0, 0, 1, 0, 1, 0, 2, // nrom                                                [sram r/w]
  1, 1, 5, 0, 5, 0, 3, // mmc1                                                [sram r/w]
  2, 3, 4, 0, 0, 0, 0, // uxrom
  3, 0, 1, 0, 3, 0, 0, // cnrom
  4, 1, 5, 0, 6, 0, 1, // mmc3/mmc6                                           [sram/prgram r/w]
  5, 3, 5, 5, 7, 0, 3, // mmc5                                                [sram r/w]
  7, 3, 4, 0, 0, 0, 0, // axrom
  9, 3, 3, 5, 5, 0, 0, // mmc2 (punch out)
````

After:
````
******************************************
  Supported Mappers
 *****************************************/
// Supported Mapper Array (iNES Mapper #s)
// Format = {mapper,prglo,prghi,chrlo,chrhi,ramlo,ramhi}
static const byte PROGMEM mapsize [] = {
  0, 0, 1, 0, 1, 0, 2, // nrom                                                [sram r/w]
  1, 1, 5, 0, 5, 0, 3, // mmc1                                                [sram r/w]
  2, 3, 4, 0, 0, 0, 0, // uxrom
  3, 0, 1, 0, 3, 0, 0, // cnrom
  4, 1, 5, 0, 6, 0, 1, // mmc3/mmc6                                           [sram/prgram r/w]
  5, 3, 5, 5, 7, 0, 3, // mmc5                                                [sram r/w]
  7, 2, 4, 0, 0, 0, 0, // axrom
  9, 3, 3, 5, 5, 0, 0, // mmc2 (punch out)
````

For reference:
````
int PRG[] = {16, 32, 64, 128, 256, 512};
byte prglo = 0; // Lowest Entry
byte prghi = 5; // Highest Entry

int CHR[] = {0, 8, 16, 32, 64, 128, 256, 512};
byte chrlo = 0; // Lowest Entry
byte chrhi = 7; // Highest Entry

byte RAM[] = {0, 8, 16, 32};
byte ramlo = 0; // Lowest Entry
byte ramhi = 3; // Highest Entry
````

**Comment by whitezombie2000** [comment]
Confirmed and working after changing 3 to 2 in NES.ino and re-flashing v8.4. I can now choose 64, 128, and 256. I tested R.C. Pro-Am and it dumped fine.

However, it did not create a CART.nes automatically with a header. It created a CART.bin headerless ROM. I manually added a header and renamed the file the CART.nes for testing. I tested Battletoads, another iNES mapper 7 game, and it also created a CART.bin headerless ROM. These must be missing from the database or the Cart Reader has some issue automatically adding the header.

**Comment by sanni** [comment]
If you don't mind experimenting you can try the current 9.0 Beta code. It comes with a whole new set of database files, so you need to replace them on your SD and mark them as hidden afterwards again so they don't clutter up the file browser. 

In 9.0 Beta I have rewritten the NES database and iNES header stuff. (The old way is still in there and can be enabled again as an option in case one prefers). 

![shegrageagdga](https://user-images.githubusercontent.com/19169494/178110504-1814c2e5-e499-444c-ad18-2da6fdea1562.jpg)

The new code automatically configures the mapper (currently that only works on a few carts, since I don't know much about NES in general) and renames the file according to no-intro.

![1](https://user-images.githubusercontent.com/19169494/178111010-acd566cb-cdc2-4e7e-8ae8-468828a44f98.jpg)

![2](https://user-images.githubusercontent.com/19169494/178111011-39fcddfa-8dde-41ab-a427-9c05838d1118.jpg)

![3](https://user-images.githubusercontent.com/19169494/178111012-7026cc46-9197-4945-b79e-eccd696647e5.jpg)

**Comment by whitezombie2000** [comment]
Yes the v9.0 Beta, using Cart Reader v3, automatically configures the mapper, and creates the .nes file with the header for R.C. Pro-Am. I had to manually set the mapper and PRG for Battletoads but it created the .nes file with the header. Both were tested and working in an emulator, great work!

**Comment by ghost** [comment]
I'll have to look into the Beta, thanks for all the help so far.
